src/jit/codegenxarch.cpp

```diff
@@ -6222,10 +6222,27 @@ void CodeGen::genCompareInt(GenTreePtr treeNode)
     assert(!varTypeIsFloating(op2Type));
 
     instruction ins;
+    var_types   type = TYP_UNKNOWN;
 
     if (tree->OperIs(GT_TEST_EQ, GT_TEST_NE))
     {
         ins = INS_test;
+
+        // Unlike many xarch instructions TEST doesn't have a form with a 16/32/64 bit first operand and
+        // an 8 bit immediate second operand. But if the immediate value fits in 8 bits then we can simply
+        // emit a 8 bit TEST instruction, unless we're targeting x86 and the first operand is a non-byteable
+        // register.
+        // Note that lowering does something similar but its main purpose is to allow memory operands to be
+        // contained so it doesn't handle other kind of operands. It could do more but on x86 that results
+        // in additional register constrains and that may be worse than wasting 3 bytes on an immediate.
+        if (
+#ifdef _TARGET_X86_
+            (!op1->isUsedFromReg() || isByteReg(op1->gtRegNum)) &&
+#endif
+            (op2->IsCnsIntOrI() && genTypeCanRepresentValue(TYP_UBYTE, op2->AsIntCon()->IconValue())))
+        {
+            type = TYP_UBYTE;
+        }
```

Comment:
Think of it as testing of a mask.  If the result of the AND with the mask, in this case 0b00000110,  is zero (ZF flag set) then JE (which is an alias of JZ) branches, the NE case is just ZF == 0.    It always takes me a bit to get my head straight on this one.  Whether any bits were set in the upper part of op1 doesn't impact the result since they were going to be AND'd with constant zeros.   (I always have to remember that CMP is SUB and that's why not all CMPs can be made into TESTs)